File: sdks/dotnet/src/Autlantic.Billing/AutlanticBillingException.cs
namespace Autlantic.Billing;

/// <summary>Error from Autlantic Billing client configuration or API responses.</summary>
public sealed class AutlanticBillingException : Exception
{
    public string? Code { get; }
    public int? StatusCode { get; }
    public string? RequestId { get; }
    public object? Body { get; }

    public AutlanticBillingException(string message)
        : this(message, null, null, null, null)
    {
    }

    public AutlanticBillingException(string message, string? code)
        : this(message, code, null, null, null)
    {
    }

    public AutlanticBillingException(string message, string? code, int? statusCode)
        : this(message, code, statusCode, null, null)
    {
    }

    public AutlanticBillingException(
        string message,
        string? code,
        int? statusCode,
        string? requestId)
        : this(message, code, statusCode, requestId, null)
    {
    }

    public AutlanticBillingException(
        string message,
        string? code,
        int? statusCode,
        string? requestId,
        object? body)
        : base(message)
    {
        Code = code;
        StatusCode = statusCode;
        RequestId = requestId;
        Body = body;
    }
}

File: sdks/java/src/main/java/com/autlantic/billing/AutlanticBillingException.java
package com.autlantic.billing;

/** Error from Autlantic Billing client configuration or API responses. */
public class AutlanticBillingException extends RuntimeException {
  private final String code;
  private final Integer statusCode;
  private final String requestId;
  private final Object body;

  public AutlanticBillingException(String message) {
    this(message, null, null, null, null);
  }

  public AutlanticBillingException(String message, String code) {
    this(message, code, null, null, null);
  }

  public AutlanticBillingException(String message, String code, Integer statusCode) {
    this(message, code, statusCode, null, null);
  }

  public AutlanticBillingException(
      String message, String code, Integer statusCode, String requestId) {
    this(message, code, statusCode, requestId, null);
  }

  public AutlanticBillingException(
      String message, String code, Integer statusCode, String requestId, Object body) {
    super(message);
    this.code = code;
    this.statusCode = statusCode;
    this.requestId = requestId;
    this.body = body;
  }

  public String getCode() {
    return code;
  }

  public Integer getStatusCode() {
    return statusCode;
  }

  public String getRequestId() {
    return requestId;
  }

  public Object getBody() {
    return body;
  }
}
